libcove2/common.py

```diff
@@ -12,32 +12,132 @@
 )
 
 
-def schema_dict_fields_generator(schema_dict):
+def _resolve_ref(value, defs, registry=None):
```

Comment:
From the code I've seen, schema_dict_fields_generator is passed a schema on which jsonref.replace_refs (or equivalent) has already been called. jsonref's default loader doesn't support URNs, but this support can be added by using a custom [loader](https://jsonref.readthedocs.io/en/latest/#jsonref.replace_refs) instead of the default [jsonloader](https://jsonref.readthedocs.io/en/latest/#jsonref.jsonloader). It would be cleaner to resolve the refs outside schema_dict_fields_generator, rather than giving schema_dict_fields_generator the responsibility for cleaning up the refs left unresolved by the default loader.

Reply:
Me and @radix0000 have had a few chats about this before and we just discussed it again. 0.2 release of this library has been made, so it is part of functionality for now (and we have a release that's dependent on this).  

But we will research some questions we have about schema wrangling before we add more like this to this library. I'd also like to get to a place where we can pass a fully de-referenced schema - not just for schema_dict_fields_generator, but for all the other functions CoVE has that might use a schema. If we can sort that out we'll unpick this work by clearly documenting expectations and marking the relevant parameters deprecated before removing them later.

Reply:
FYI, for the OCDS Data Review Tool, I’m switching to using get_schema_fields in OCDS Kit for all schema analysis. This originally was only for the mapping sheet. https://github.com/open-contracting/ocdskit/blob/c3f341b1dd3d941a3ac4443203d7e271509fae22/ocdskit/schema.py#L66

It takes a deferenced schema and extracts all metadata (deprecated, required etc) in a single pass. There are tests (test_regression) to check that it can produce the same results as libcove with equal or better running time. As far as I know, it supports BODS except for URN support (but that support can be added to the jsonref loader).

The tests in test_regression make a few notes of bugs in libcove (missing support for oneOf in the deprecated and required ID functions, etc.).